ElectricCar.cpp

```diff
@@ -1,16 +1,28 @@
 #include "ElectricCar.hpp"
 #include <iostream>
+#include "CarInMovementExcept.hpp"
 
 ElectricCar::ElectricCar(ElectricEngine* engine)
-    : engine_(engine)
-{
-    std::cout << __FUNCTION__ << std::endl;
+    : electricEngine_(engine) {
 }
 
-ElectricCar::~ElectricCar()       { std::cout << __FUNCTION__ << std::endl; }
-void ElectricCar::turnLeft()      { std::cout << __FUNCTION__ << std::endl; }
-void ElectricCar::turnRight()     { std::cout << __FUNCTION__ << std::endl; }
-void ElectricCar::brake()         { std::cout << __FUNCTION__ << std::endl; }
-void ElectricCar::accelerate(int) { std::cout << __FUNCTION__ << std::endl; }
-void ElectricCar::charge()        { std::cout << __FUNCTION__ << std::endl; }
+ElectricCar::~ElectricCar() {
+}
+void ElectricCar::refill() {
+    charge();
+}
+void ElectricCar::charge() {
+}
 
+void ElectricCar::changeEngine(Engine* electricEngine) {
+    if (currentSpeed_ != 0) {
+        throw CarInMovementExcept("Car is moving");
+    } else if (typeid(*electricEngine) == typeid(ElectricEngine)) {
+        delete electricEngine_;
+        electricEngine_ = static_cast<ElectricEngine*>(electricEngine);
```

Comment:
Isn't better to just take ElectricEngine as car parameter?

Reply:
No because change engine is derived from Car and it has to engine type independent